sys/kern/proc.c

```diff
@@ -253,21 +288,19 @@ __noreturn void proc_exit(int exitstatus) {
   proc_unlock(p);
 
   WITH_MTX_LOCK (all_proc_mtx) {
+    if (p->p_pid == 1)
+      panic("'init' process died!");
+
     /* Process orphans, but firstly find init process. */
-    proc_t *init;
-    TAILQ_FOREACH (init, &proc_list, p_all) {
-      if (init->p_pid == 1)
-        break;
-    }
+    proc_t *init = proc_find_raw(1);
+    assert(init != NULL);
     proc_reparent(p, init);
 
     TAILQ_REMOVE(&proc_list, p, p_all);
     TAILQ_INSERT_TAIL(&zombie_list, p, p_zombie);
 
     /* When the process is dead we can finally signal the parent. */
     proc_t *parent = p->p_parent;
-    if (!parent)
```

Comment:
To remove all checks for `p->p_parent` being `NULL` we should finally introduce `proc0` as they do in [FreeBSD](http://bxr.su/FreeBSD/sys/kern/init_main.c#proc0_init) and [NetBSD](http://bxr.su/NetBSD/sys/kern/kern_proc.c#proc0_init). What do you think?

Reply:
I think it's a good idea, but someone needs to implement it.

Reply:
You're right, it's better now to focus on adding sessions to the system.